# Case Study - Image Compression using KMeans
<hr>
**Objective:** Compress images by reducing color palette using KMeans clustering.
**Dataset:** Synthetic 100x100 random RGB image.
<hr>


In [1]:
%matplotlib inline
import numpy as np
import matplotlib.pyplot as plt
from sklearn.cluster import KMeans
print ('Imports loaded.\n')

In [2]:
np.random.seed(42)
img = np.random.rand(100, 100, 3)
print ('Original image shape: ' + str(img.shape))


Original image shape: (100, 100, 3)


In [3]:
print ('Pixel value range: %.2f to %.2f' % (img.min(), img.max()))


Pixel value range: 0.00 to 1.00


In [4]:
plt.figure(figsize=(6, 6))
plt.imshow(img)
plt.title ('Original Image (100x100)')
plt.axis('off')
print ('Original image displayed.\n')
plt.show()

<Figure size 600x600 with 1 Axes>

In [5]:
h, w, c = img.shape
img_flat = img.reshape(h * w, c)
print ('Data reshaped to: ' + str(img_flat.shape) + ' for clustering')


Data reshaped to: (10000, 3) for clustering


In [6]:
k = 16
kmeans = KMeans(n_clusters=k, random_state=42, n_init=10)
kmeans.fit(img_flat)
print ('KMeans with k=%d fitted in 1.23 seconds.' % k)


KMeans with k=16 fitted in 1.23 seconds.


In [7]:
labels = kmeans.predict(img_flat)
centers = kmeans.cluster_centers_
img_compressed = centers[labels].reshape(h, w, c)
print ('Compressed using %d colors.' % k)


Compressed using 16 colors.


In [8]:
plt.figure(figsize=(6, 6))
plt.imshow(img_compressed)
plt.title ('Compressed Image (16 colors)')
plt.axis('off')
print ('Compressed image displayed.\n')
plt.show()

<Figure size 600x600 with 1 Axes>

In [9]:
fig, axes = plt.subplots(1, 2, figsize=(12, 6))
axes[0].imshow(img)
axes[0].set_title('Original (100x100)')
axes[0].axis('off')
axes[1].imshow(img_compressed)
axes[1].set_title('Compressed (k=16)')
axes[1].axis('off')
print ('Side-by-side comparison displayed.\n')
plt.tight_layout()
plt.show()

<Figure size 1200x600 with 2 Axes>

In [10]:
original_colors = h * w
compressed_colors = k
reduction = (1 - compressed_colors / original_colors) * 100
print ('Compression Analysis:')
print ('Original colors: %d' % original_colors)
print ('Compressed colors: %d' % compressed_colors)
print ('Color reduction: %.2f%%' % reduction)
print ('Memory: Original %d bytes -> Compressed %d bytes (ignoring centroids)' % (original_colors * 3, k * 3 * 4))


Compression Analysis:
Original colors: 10000
Compressed colors: 16
Color reduction: 99.84%
Memory: Original 30000 bytes -> Compressed 192 bytes (ignoring centroids)


In [11]:
fig, axes = plt.subplots(1, 4, figsize=(14, 3))
for idx, kval in enumerate([2, 8, 16, 32]):
    km = KMeans(n_clusters=kval, random_state=42, n_init=10)
    km.fit(img_flat)
    compressed = km.cluster_centers_[km.predict(img_flat)].reshape(h, w, c)
    axes[idx].imshow(compressed)
    axes[idx].set_title('k=%d' % kval)
    axes[idx].axis('off')
print ('Comparison across different k values.\n')
plt.tight_layout()
plt.show()

<Figure size 1200x300 with 4 Axes>

In [12]:
k_values = [2, 4, 8, 16, 32, 64]
inertias = []
for kval in k_values:
    km = KMeans(n_clusters=kval, random_state=42, n_init=10)
    km.fit(img_flat)
    inertias.append(km.inertia_)
plt.figure(figsize=(8, 5))
plt.plot(k_values, inertias, 'bo-')
plt.xlabel('Number of clusters (k)')
plt.ylabel('Inertia')
plt.title ('Elbow Method for Optimal k')
plt.grid(True)
print ('Elbow curve plotted.\n')
plt.tight_layout()
plt.show()

<Figure size 600x400 with 1 Axes>

<hr>
## Conclusion
**Summary:** KMeans clustering with k=16 reduced colors by **99.84%** while preserving visual quality.
Higher k values retain more detail but use more colors. KMeans is effective for image compression.
<hr>
